File: pfm-sdk/src/main/java/com/tink/pfmsdk/view/TinkSnackbar.java
package com.tink.pfmsdk.view;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewGroup.LayoutParams;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.snackbar.SnackbarContentLayout;
import com.tink.pfmsdk.R;
import com.tink.pfmsdk.util.DimensionUtils;

public class TinkSnackbar {

	private static final float LOADER_SIZE_DP = 35;
	private static final float ICON_SIZE_DP = 35;

	private SnackbarContentLayout snackbarContentLayout;

	private Snackbar wrappedSnackbar;

	private Context context;

	private ProgressBar loadingIndicator;

	private Theme theme;

	private TinkSnackbar(Snackbar snackbar, Theme theme) {
		this.wrappedSnackbar = snackbar;
		this.context = snackbar.getContext();
		this.theme = theme;
		snackbarContentLayout = findRecursive(wrappedSnackbar.getView());

		wrappedSnackbar.getView().setBackgroundColor(theme.getBackgroundColor());

		TextView messageView = snackbarContentLayout.findViewById(R.id.snackbar_text);
		TextView actionView = snackbarContentLayout.findViewById(R.id.snackbar_action);

		styleText(messageView, theme.getTextTheme());
		styleText(actionView, theme.getButtonTheme());
	}

	public boolean isShown() {
		return wrappedSnackbar.isShown();
	}

	public static TinkSnackbar make(View view, String text, int duration, Theme theme) {
		Snackbar snackbar = Snackbar.make(view, text, duration);
		LayoutParams lp = snackbar.getView().getLayoutParams();

		// TODO: PFMSDK: How can he handle this with the app's bottom navigation view?
		/*boolean isBottomNavigationBar = view.findViewById(R.id.bottomNavigation) != null
			&& lp instanceof CoordinatorLayout.LayoutParams;
		if (isBottomNavigationBar) {
			CoordinatorLayout.LayoutParams clp = (CoordinatorLayout.LayoutParams) lp;
			clp.setAnchorId(R.id.bottomNavigation);
			clp.anchorGravity = Gravity.TOP;
			clp.gravity = Gravity.TOP;
			snackbar.getView().setLayoutParams(clp);
		}*/
		TinkSnackbar instance = new TinkSnackbar(snackbar, theme);
		View snackbarView = findRecursive(snackbar.getView());
		if (snackbarView != null) {
			snackbarView
				.setMinimumHeight((int) view.getResources().getDimension(R.dimen.snackbar_height));
		}
		return instance;
	}

	public void setLoading(boolean loading) {
		if (!loading) {
			if (loadingIndicator != null) {
				loadingIndicator.setVisibility(View.GONE);
			}
		} else {
			if (loadingIndicator == null) {
				initLoader();
			}
			loadingIndicator.setVisibility(View.VISIBLE);
		}
	}

	public void setAction(String text, View.OnClickListener listener) {
		wrappedSnackbar.setAction(text, listener);
	}

	public void setIcon(Drawable drawable) {
		ImageView icon = new ImageView(context);
		icon.setImageDrawable(drawable);
		int iconSize = Math.round(DimensionUtils.getPixelsFromDP(ICON_SIZE_DP, context));
		LinearLayout.LayoutParams iconLayoutParams = new LinearLayout.LayoutParams(
			new LayoutParams(iconSize, iconSize));
		iconLayoutParams.gravity = Gravity.CENTER_VERTICAL;
		icon.setLayoutParams(iconLayoutParams);
		snackbarContentLayout.addView(icon);
	}

	public void show() {
		wrappedSnackbar.show();
	}

	public Snackbar getWrappedSnackbar() {
		return wrappedSnackbar;
	}

	private static SnackbarContentLayout findRecursive(View root) {
		if (root instanceof SnackbarContentLayout) {
			return (SnackbarContentLayout) root;
		} else if (root instanceof ViewGroup) {
			ViewGroup viewGroup = (ViewGroup) root;
			for (int i = 0; i < viewGroup.getChildCount(); i++) {
				View child = viewGroup.getChildAt(i);
				SnackbarContentLayout foundInChildren = findRecursive(child);
				if (foundInChildren != null) {
					return foundInChildren;
				}
			}
		}
		return null;
	}

	private static void styleText(TextView textView, TinkTextView.Theme theme) {
		textView.setTypeface(theme.getFont());
		textView.setTextColor(theme.getTextColor());
	}

	private void initLoader() {
		loadingIndicator = new ProgressBar(context);
		int progressSize = Math.round(DimensionUtils.getPixelsFromDP(LOADER_SIZE_DP, context));
		LinearLayout.LayoutParams progressLayoutParams = new LinearLayout.LayoutParams(
			new LayoutParams(progressSize, progressSize));
		progressLayoutParams.gravity = Gravity.CENTER_VERTICAL;
		loadingIndicator.setLayoutParams(progressLayoutParams);
		loadingIndicator.getIndeterminateDrawable().setColorFilter(theme.getLoadingIndicatorColor(),
			android.graphics.PorterDuff.Mode.SRC_IN);
		loadingIndicator.setVisibility(View.GONE);
		snackbarContentLayout.addView(loadingIndicator);
	}

	public void dismiss() {
		wrappedSnackbar.dismiss();
	}

	public interface Theme {

		String MESSAGE_THEME = "message_theme";
		String ERROR_THEME = "error_theme";

		TinkTextView.Theme getTextTheme();

		TinkTextView.Theme getButtonTheme();

		int getBackgroundColor();

		int getLoadingIndicatorColor();
	}
}

File: pfm-sdk/src/main/java/com/tink/pfmsdk/view/TinkSnackbar.kt
package com.tink.pfmsdk.view

import android.content.Context
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.snackbar.SnackbarContentLayout
import com.tink.pfmsdk.R
import com.tink.pfmsdk.util.DimensionUtils
import kotlin.math.roundToInt

internal class TinkSnackbar private constructor(
    private val wrappedSnackbar: Snackbar,
    private val theme: Theme
) {
    private val snackbarContentLayout: SnackbarContentLayout?
    private val context: Context = wrappedSnackbar.context
    private var loadingIndicator: ProgressBar? = null

    init {
        snackbarContentLayout = findRecursive(wrappedSnackbar.view)
        wrappedSnackbar.view.setBackgroundColor(theme.backgroundColor)
        val messageView =
            snackbarContentLayout!!.findViewById<TextView>(R.id.snackbar_text)
        val actionView =
            snackbarContentLayout.findViewById<TextView>(R.id.snackbar_action)
        styleText(messageView, theme.textTheme)
        styleText(actionView, theme.buttonTheme)
    }

    fun setLoading(loading: Boolean) {
        if (!loading) {
            if (loadingIndicator != null) {
                loadingIndicator!!.visibility = View.GONE
            }
        } else {
            if (loadingIndicator == null) {
                initLoader()
            }
            loadingIndicator!!.visibility = View.VISIBLE
        }
    }

    fun setAction(text: String?, listener: View.OnClickListener?) {
        wrappedSnackbar.setAction(text, listener)
    }

    fun setIcon(drawable: Drawable?) {
        val icon = ImageView(context)
        icon.setImageDrawable(drawable)
        val iconSize = Math.round(
            DimensionUtils.getPixelsFromDP(
                ICON_SIZE_DP,
                context
            )
        )
        val iconLayoutParams =
            LinearLayout.LayoutParams(
                ViewGroup.LayoutParams(iconSize, iconSize)
            )
        iconLayoutParams.gravity = Gravity.CENTER_VERTICAL
        icon.layoutParams = iconLayoutParams
        snackbarContentLayout!!.addView(icon)
    }

    fun show() {
        wrappedSnackbar.show()
    }

    private fun initLoader() {
        loadingIndicator = ProgressBar(context)
        val progressSize = DimensionUtils.getPixelsFromDP(
            LOADER_SIZE_DP,
            context
        ).roundToInt()
        val progressLayoutParams =
            LinearLayout.LayoutParams(
                ViewGroup.LayoutParams(progressSize, progressSize)
            )
        progressLayoutParams.gravity = Gravity.CENTER_VERTICAL
        loadingIndicator!!.layoutParams = progressLayoutParams
        loadingIndicator!!.indeterminateDrawable.setColorFilter(
            theme.loadingIndicatorColor,
            PorterDuff.Mode.SRC_IN
        )
        loadingIndicator!!.visibility = View.GONE
        snackbarContentLayout!!.addView(loadingIndicator)
    }

    fun dismiss() {
        wrappedSnackbar.dismiss()
    }

    interface Theme {
        val textTheme: TinkTextView.Theme
        val buttonTheme: TinkTextView.Theme
        val backgroundColor: Int
        val loadingIndicatorColor: Int

        companion object {
            const val MESSAGE_THEME = "message_theme"
            const val ERROR_THEME = "error_theme"
        }
    }

    companion object {
        private const val LOADER_SIZE_DP = 35f
        private const val ICON_SIZE_DP = 35f
        fun make(
            view: View,
            text: String?,
            duration: Int,
            theme: Theme
        ): TinkSnackbar {
            val snackbar = Snackbar.make(view, text!!, duration)
            // TODO: PFMSDK: How can he handle this with the app's bottom navigation view?
            /*val lp = snackbar.view.layoutParams
            boolean isBottomNavigationBar = view.findViewById(R.id.bottomNavigation) != null
			&& lp instanceof CoordinatorLayout.LayoutParams;
		    if (isBottomNavigationBar) {
                CoordinatorLayout.LayoutParams clp = (CoordinatorLayout.LayoutParams) lp;
                clp.setAnchorId(R.id.bottomNavigation);
                clp.anchorGravity = Gravity.TOP;
                clp.gravity = Gravity.TOP;
                snackbar.getView().setLayoutParams(clp);
		    }*/
            val instance = TinkSnackbar(snackbar, theme)
            val snackbarView: View? =
                findRecursive(snackbar.view)
            if (snackbarView != null) {
                snackbarView.minimumHeight =
                    view.resources.getDimension(R.dimen.snackbar_height).toInt()
            }
            return instance
        }

        private fun findRecursive(root: View): SnackbarContentLayout? {
            if (root is SnackbarContentLayout) {
                return root
            } else if (root is ViewGroup) {
                for (i in 0 until root.childCount) {
                    val child = root.getChildAt(i)
                    val foundInChildren =
                        findRecursive(child)
                    if (foundInChildren != null) {
                        return foundInChildren
                    }
                }
            }
            return null
        }

        private fun styleText(
            textView: TextView,
            theme: TinkTextView.Theme
        ) {
            textView.typeface = theme.font
            textView.setTextColor(theme.textColor)
        }
    }
}